test/cqlpy/test_vector_index.py

```diff
@@ -337,3 +337,141 @@ def test_vector_search_when_tracing_is_enabled(cql, test_keyspace, scylla_only):
+        with pytest.raises(InvalidRequest, match=f"Function system.similarity_{similarity_function} requires a float vector argument" if is_scylla(cql) else "Type error"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}(v, [0.3, '2025-10-07T14:05:00.000', 20.25]) FROM {table} ORDER BY v ANN OF [0.3, 1.0, 20.25] LIMIT 1")
+        with pytest.raises(InvalidRequest, match=f"Function system.similarity_{similarity_function} requires a float vector argument" if is_scylla(cql) else "Type error"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}(v, [0.3, 'dog', 20.25]) FROM {table} ORDER BY v ANN OF [0.3, 1.0, 20.25] LIMIT 1")
+        with pytest.raises(InvalidRequest, match=f"Function system.similarity_{similarity_function} requires a float vector argument" if is_scylla(cql) else "Type error"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}(v, [0.3, {{1.0, 2.0, 3.0}}, 20.25]) FROM {table} ORDER BY v ANN OF [0.3, 1.0, 20.25] LIMIT 1")
+        # This test is Scylla-only because Cassandra does not handle it properly and crashes on org.apache.cassandra.serializers.MarshalException.
+        if is_scylla(cql):
+            with pytest.raises(InvalidRequest, match=f"null is not supported inside vectors"):
+                cql.execute(f"SELECT pk, similarity_{similarity_function}(v, [0.3, null, 20.25]) FROM {table} ORDER BY v ANN OF [0.3, 1.0, 20.25] LIMIT 1")
+        cql.execute(f"DROP INDEX {test_keyspace}.{table.split('.')[1]}_v_idx")
+
+
+# The test is Scylla-only because Cassandra does not require the similarity function to match the index configuration.
+@pytest.mark.parametrize("similarity_function", ["cosine", "euclidean", "dot_product"])
+def test_vector_similarity_fails_on_mismatched_function(cql, test_keyspace, scylla_only, similarity_function):
+    schema = 'pk int primary key, v vector<float, 3>'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (pk, v) VALUES (1, [1.0, 2.0, 3.0])")
+        cql.execute(f"CREATE CUSTOM INDEX ON {table}(v) USING 'vector_index' WITH OPTIONS = {{'similarity_function': '{similarity_function}'}}")
+
+        # Only the configured similarity function should succeed; others should fail.
+        for func in ["cosine", "euclidean", "dot_product"]:
+            query = f"SELECT pk, similarity_{func}(v, [0.3, 1.0, 20.25]) FROM {table} ORDER BY v ANN OF [0.3, 1.0, 20.25] LIMIT 1"
+            if func != similarity_function:
+                with pytest.raises(InvalidRequest, match=f"Vector index on column v was not built with similarity_{func} metric"):
+                    cql.execute(query)
+        cql.execute(f"DROP INDEX {test_keyspace}.{table.split('.')[1]}_v_idx")
+
+
+# The test is Scylla-only because Cassandra does not require the arguments to be the same as for the ANN query.
```

Comment:
Do you know why Cassandra doesn't require this?
Again it's not clear to me if "scylla_only" is the right verdict on this test - or if it should be xfail (a Scylla bug) or cassandra_bug (if it's a Cassandra bug), or what...